## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [55]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df[df["order_id"] == 2742071]

first   last                                       job  \
3446   Kimberly  Blair  Geographical information systems officer   
6287   Kimberly  Blair  Geographical information systems officer   
20623  Kimberly  Blair  Geographical information systems officer   
23840  Kimberly  Blair  Geographical information systems officer   
24671  Kimberly  Blair  Geographical information systems officer   
42470  Kimberly  Blair  Geographical information systems officer   

                    phone                           email  client_id  \
3446   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
6287   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
20623  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
23840  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
24671  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
42470  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   

       order_id  order_date  order_week  order_year         item_id  \
3446    2742071  2023-02-27           9        2023  ACU90518-83-2B   
6287    2742071  2023-02-27           9        2023  BMD72525-90-2X   
20623   2742071  2023-02-27           9        2023  DEE52027-40-3U   
23840   2742071  2023-02-27           9        2023  AEC69505-92-4M   
24671   2742071  2023-02-27           9        2023  AEE28845-01-5Z   
42470   2742071  2023-02-27           9        2023  BDB35685-37-8X   

          category         subcategory  unit_price  unit_cost  unit_weight  \
3446   consumables   bathroom supplies        5.59       3.05        10.46   
6287   consumables   bathroom supplies      460.20     407.68         2.60   
20623     software  project management      135.12      80.78         2.27   
23840    furniture              tables       47.59      35.41         0.54   
24671  consumables    kitchen supplies       59.69      54.86         1.57   
42470        decor            wall art       82.18      61.07         3.04   

       qty  line_number  
3446    77            5  
6287   260            0  
20623   12            3  
23840   22            1  
24671   24            2  
42470   44            4

In [63]:
# View the column names in the data
df.columns.tolist()


['first',
 'last',
 'job',
 'phone',
 'email',
 'client_id',
 'order_id',
 'order_date',
 'order_week',
 'order_year',
 'item_id',
 'category',
 'subcategory',
 'unit_price',
 'unit_cost',
 'unit_weight',
 'qty',
 'line_number']

In [61]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [38]:
# Use this space to do any additional research
# and familiarize yourself with the data.
# df['item_id'].describe()
item_qty_sum = df.groupby('item_id')['qty'].sum()
item_qty_sum.sort_values(ascending=False).head(4)


item_id
XAE75138-91-2E    3998066
CCU49971-42-9U     994154
XDA18116-89-4A     895640
XXM82450-88-9M     670190
Name: qty, dtype: int64

In [76]:
# What three item categories had the most entries?
df['category'].value_counts(ascending=False).head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [109]:
# For the category with the most entries, which subcategory had the most entries?
df[ ['category', 'subcategory'] ].value_counts(ascending=False).head(1)


category     subcategory      
consumables  bathroom supplies    6424
Name: count, dtype: int64

In [41]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts(ascending=False).head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [81]:
# Store the client ids of those top 5 clients in a list.
top_5_client_list = df['client_id'].value_counts(ascending=False).head(5).index.tolist()
print(top_5_client_list)

[33615, 66037, 46820, 38378, 24741]


In [88]:
# How many total units (the qty column) did the client with the most entries order order?
units_total = df.loc[df['client_id'] == 33615, 'qty'].sum()
print(units_total)
# df[df['client_id'] == 33615].groupby('client_id')['qty'].sum()

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [98]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df[ ['unit_price', 'qty', 'subtotal'] ].head(2)
# df[df['order_id'] == 2742071]

first   last                                       job  \
3446   Kimberly  Blair  Geographical information systems officer   
6287   Kimberly  Blair  Geographical information systems officer   
20623  Kimberly  Blair  Geographical information systems officer   
23840  Kimberly  Blair  Geographical information systems officer   
24671  Kimberly  Blair  Geographical information systems officer   
42470  Kimberly  Blair  Geographical information systems officer   

                    phone                           email  client_id  \
3446   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
6287   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
20623  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
23840  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
24671  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
42470  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   

       order_id  order_date  order_week  order_year         item_id  \
3446    2742071  2023-02-27           9        2023  ACU90518-83-2B   
6287    2742071  2023-02-27           9        2023  BMD72525-90-2X   
20623   2742071  2023-02-27           9        2023  DEE52027-40-3U   
23840   2742071  2023-02-27           9        2023  AEC69505-92-4M   
24671   2742071  2023-02-27           9        2023  AEE28845-01-5Z   
42470   2742071  2023-02-27           9        2023  BDB35685-37-8X   

          category         subcategory  unit_price  unit_cost  unit_weight  \
3446   consumables   bathroom supplies        5.59       3.05        10.46   
6287   consumables   bathroom supplies      460.20     407.68         2.60   
20623     software  project management      135.12      80.78         2.27   
23840    furniture              tables       47.59      35.41         0.54   
24671  consumables    kitchen supplies       59.69      54.86         1.57   
42470        decor            wall art       82.18      61.07         3.04   

       qty  line_number   subtotal  
3446    77            5     430.43  
6287   260            0  119652.00  
20623   12            3    1621.44  
23840   22            1    1046.98  
24671   24            2    1432.56  
42470   44            4    3615.92

In [45]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['order_weight'] = df['unit_weight'] * df['qty']
df['shipping_rate'] = df.apply(lambda row: 7 if row['order_weight'] > 50 else 10, axis=1)
df['shipping_total'] = df['unit_weight'] * df['qty'] * df['shipping_rate'] 
df[df["order_id"] == 2742071]

first   last                                       job  \
3446   Kimberly  Blair  Geographical information systems officer   
6287   Kimberly  Blair  Geographical information systems officer   
20623  Kimberly  Blair  Geographical information systems officer   
23840  Kimberly  Blair  Geographical information systems officer   
24671  Kimberly  Blair  Geographical information systems officer   
42470  Kimberly  Blair  Geographical information systems officer   

                    phone                           email  client_id  \
3446   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
6287   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
20623  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
23840  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
24671  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
42470  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   

       order_id  order_date  order_week  order_year  ...         subcategory  \
3446    2742071  2023-02-27           9        2023  ...   bathroom supplies   
6287    2742071  2023-02-27           9        2023  ...   bathroom supplies   
20623   2742071  2023-02-27           9        2023  ...  project management   
23840   2742071  2023-02-27           9        2023  ...              tables   
24671   2742071  2023-02-27           9        2023  ...    kitchen supplies   
42470   2742071  2023-02-27           9        2023  ...            wall art   

      unit_price unit_cost  unit_weight  qty  line_number   subtotal  \
3446        5.59      3.05        10.46   77            5     430.43   
6287      460.20    407.68         2.60  260            0  119652.00   
20623     135.12     80.78         2.27   12            3    1621.44   
23840      47.59     35.41         0.54   22            1    1046.98   
24671      59.69     54.86         1.57   24            2    1432.56   
42470      82.18     61.07         3.04   44            4    3615.92   

       order_weight  shipping_rate  shipping_total  
3446         805.42              7         5637.94  
6287         676.00              7         4732.00  
20623         27.24             10          272.40  
23840         11.88             10          118.80  
24671         37.68             10          376.80  
42470        133.76              7          936.32  

[6 rows x 22 columns]

In [46]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['charged_price_pre_tax'] = df['subtotal'] + df['shipping_total']
df['tax'] = df['charged_price_pre_tax'] * 0.0925
df['charged_price'] = df['tax'] + df['charged_price_pre_tax']
df[df["order_id"] == 2742071]

first   last                                       job  \
3446   Kimberly  Blair  Geographical information systems officer   
6287   Kimberly  Blair  Geographical information systems officer   
20623  Kimberly  Blair  Geographical information systems officer   
23840  Kimberly  Blair  Geographical information systems officer   
24671  Kimberly  Blair  Geographical information systems officer   
42470  Kimberly  Blair  Geographical information systems officer   

                    phone                           email  client_id  \
3446   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
6287   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
20623  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
23840  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
24671  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
42470  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   

       order_id  order_date  order_week  order_year  ... unit_weight  qty  \
3446    2742071  2023-02-27           9        2023  ...       10.46   77   
6287    2742071  2023-02-27           9        2023  ...        2.60  260   
20623   2742071  2023-02-27           9        2023  ...        2.27   12   
23840   2742071  2023-02-27           9        2023  ...        0.54   22   
24671   2742071  2023-02-27           9        2023  ...        1.57   24   
42470   2742071  2023-02-27           9        2023  ...        3.04   44   

      line_number   subtotal  order_weight  shipping_rate  shipping_total  \
3446            5     430.43        805.42              7         5637.94   
6287            0  119652.00        676.00              7         4732.00   
20623           3    1621.44         27.24             10          272.40   
23840           1    1046.98         11.88             10          118.80   
24671           2    1432.56         37.68             10          376.80   
42470           4    3615.92        133.76              7          936.32   

       charged_price_pre_tax           tax  charged_price  
3446                 6068.37    561.324225    6629.694225  
6287               124384.00  11505.520000  135889.520000  
20623                1893.84    175.180200    2069.020200  
23840                1165.78    107.834650    1273.614650  
24671                1809.36    167.365800    1976.725800  
42470                4552.24    421.082200    4973.322200  

[6 rows x 25 columns]

In [47]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['mnfg_cost'] = df['unit_cost'] * df['qty'] + df['shipping_total'] 
df[df["order_id"] == 2742071]


first   last                                       job  \
3446   Kimberly  Blair  Geographical information systems officer   
6287   Kimberly  Blair  Geographical information systems officer   
20623  Kimberly  Blair  Geographical information systems officer   
23840  Kimberly  Blair  Geographical information systems officer   
24671  Kimberly  Blair  Geographical information systems officer   
42470  Kimberly  Blair  Geographical information systems officer   

                    phone                           email  client_id  \
3446   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
6287   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
20623  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
23840  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
24671  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
42470  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   

       order_id  order_date  order_week  order_year  ...  qty line_number  \
3446    2742071  2023-02-27           9        2023  ...   77           5   
6287    2742071  2023-02-27           9        2023  ...  260           0   
20623   2742071  2023-02-27           9        2023  ...   12           3   
23840   2742071  2023-02-27           9        2023  ...   22           1   
24671   2742071  2023-02-27           9        2023  ...   24           2   
42470   2742071  2023-02-27           9        2023  ...   44           4   

        subtotal  order_weight  shipping_rate  shipping_total  \
3446      430.43        805.42              7         5637.94   
6287   119652.00        676.00              7         4732.00   
20623    1621.44         27.24             10          272.40   
23840    1046.98         11.88             10          118.80   
24671    1432.56         37.68             10          376.80   
42470    3615.92        133.76              7          936.32   

       charged_price_pre_tax           tax  charged_price  mnfg_cost  
3446                 6068.37    561.324225    6629.694225    5872.79  
6287               124384.00  11505.520000  135889.520000  110728.80  
20623                1893.84    175.180200    2069.020200    1241.76  
23840                1165.78    107.834650    1273.614650     897.82  
24671                1809.36    167.365800    1976.725800    1693.44  
42470                4552.24    421.082200    4973.322200    3623.40  

[6 rows x 26 columns]

In [48]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['charged_price'] - df['mnfg_cost']
df[df["order_id"] == 2742071]


first   last                                       job  \
3446   Kimberly  Blair  Geographical information systems officer   
6287   Kimberly  Blair  Geographical information systems officer   
20623  Kimberly  Blair  Geographical information systems officer   
23840  Kimberly  Blair  Geographical information systems officer   
24671  Kimberly  Blair  Geographical information systems officer   
42470  Kimberly  Blair  Geographical information systems officer   

                    phone                           email  client_id  \
3446   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
6287   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
20623  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
23840  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
24671  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
42470  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   

       order_id  order_date  order_week  order_year  ... line_number  \
3446    2742071  2023-02-27           9        2023  ...           5   
6287    2742071  2023-02-27           9        2023  ...           0   
20623   2742071  2023-02-27           9        2023  ...           3   
23840   2742071  2023-02-27           9        2023  ...           1   
24671   2742071  2023-02-27           9        2023  ...           2   
42470   2742071  2023-02-27           9        2023  ...           4   

        subtotal order_weight  shipping_rate  shipping_total  \
3446      430.43       805.42              7         5637.94   
6287   119652.00       676.00              7         4732.00   
20623    1621.44        27.24             10          272.40   
23840    1046.98        11.88             10          118.80   
24671    1432.56        37.68             10          376.80   
42470    3615.92       133.76              7          936.32   

       charged_price_pre_tax           tax  charged_price  mnfg_cost  \
3446                 6068.37    561.324225    6629.694225    5872.79   
6287               124384.00  11505.520000  135889.520000  110728.80   
20623                1893.84    175.180200    2069.020200    1241.76   
23840                1165.78    107.834650    1273.614650     897.82   
24671                1809.36    167.365800    1976.725800    1693.44   
42470                4552.24    421.082200    4973.322200    3623.40   

             profit  
3446     756.904225  
6287   25160.720000  
20623    827.260200  
23840    375.794650  
24671    283.285800  
42470   1349.922200  

[6 rows x 27 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [49]:
# Check your work using the totals above

print("${:,.3f}".format(df[df['order_id'] == 2742071]['charged_price'].sum()))
print("${:,.3f}".format(df[df['order_id'] == 2173913]['charged_price'].sum()))
print("${:,.3f}".format(df[df['order_id'] == 6128929]['charged_price'].sum()))

$152,811.897
$162,388.719
$923,441.240


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [50]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_5_clients_by_qty_list = df.groupby('client_id')['qty'].sum().head(5).sort_values(ascending=False).index.tolist()
print(top_5_clients_by_qty_list)
top_5_clients_by_qty_df = df[df['client_id'].isin(top_5_clients_by_qty_list)]
# print(filtered_df)
results1 = top_5_clients_by_qty_df.groupby('client_id')['charged_price'].sum()
print("Top 5 by quantity\n",results1)
# The question is actually asking for top 5 clients by quantity here. 
# In part 1 it asked for top 5 clients by number of entries. THEY DON'T MATCH! 
# Therefore, I'll create totals for the two distinct groups
results2 = top_5_clients_df.groupby('client_id')['charged_price'].sum()
print("Top 5 by entries\n",results2)

[10435, 10218, 10091, 10351, 10033]
Top 5 by quantity
 client_id
10033    6.892750e+05
10091    1.825073e+06
10218    3.683120e+06
10351    1.712403e+06
10435    8.722370e+06
Name: charged_price, dtype: float64


KeyError: 'Column not found: charged_price'

In [ ]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = top_5_clients_df.groupby('client_id')[['qty', 'shipping_total', 'charged_price', 'profit']].sum().reset_index()

print(summary_df)

   client_id     qty  shipping_total  charged_price        profit
0      24741  239862      5126448.37   8.226889e+07  3.657999e+07
1      33615   64313      1828984.89   8.377309e+06  2.201995e+06
2      38378   73667      3429455.40   1.290655e+07  3.271830e+06
3      46820   75768      1601448.84   9.743794e+06  2.736603e+06
4      66037   43018      1395151.85   1.025951e+07  3.255032e+06


In [ ]:
#AGK - Needed to re-set the dataframe (in order to run this block of code multiple times)
#AGK - because we are making destructive changes to the column names:
summary_df = top_5_clients_df.groupby('client_id')[['qty', 'shipping_total', 'charged_price', 'profit']].sum().reset_index()

# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 

shipping_totals = summary_df['shipping_total']
charged_price = summary_df['charged_price']
profit = summary_df['profit']

# Define a function that converts a dollar amount to millions.
shipping_totals = summary_df['shipping_total'] / 1000000
charged_price = summary_df['charged_price'] / 1000000
profit = summary_df['profit'] / 1000000

# Apply the currency_format_millions function to only the money columns. 
shipping_totals = shipping_totals.map("${:,.1f}M".format)
summary_df['shipping_total'] = shipping_totals

charged_price = charged_price.map("${:,.1f}M".format)
summary_df['charged_price'] = charged_price

profit = profit.map("${:,.1f}M".format)
summary_df['profit'] = profit

# Rename the columns to reflect the change in the money format. 
summary_df = summary_df.rename(columns={'shipping_total': 'Total Shiping (millions)'})
summary_df = summary_df.rename(columns={'charged_price': 'Charged Price (millions)'})
summary_df = summary_df.rename(columns={'profit': 'Total Profit (millions)'})
print(summary_df)

   client_id     qty Total Shiping (millions) Charged Price (millions)  \
0      24741  239862                    $5.1M                   $82.3M   
1      33615   64313                    $1.8M                    $8.4M   
2      38378   73667                    $3.4M                   $12.9M   
3      46820   75768                    $1.6M                    $9.7M   
4      66037   43018                    $1.4M                   $10.3M   

  Total Profit (millions)  
0                  $36.6M  
1                   $2.2M  
2                   $3.3M  
3                   $2.7M  
4                   $3.3M  


In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
summary_df.sort_values(by='Total Profit (millions)', ascending=False)

client_id     qty Total Shiping (millions) Charged Price (millions)  \
0      24741  239862                    $5.1M                   $82.3M   
2      38378   73667                    $3.4M                   $12.9M   
4      66037   43018                    $1.4M                   $10.3M   
3      46820   75768                    $1.6M                    $9.7M   
1      33615   64313                    $1.8M                    $8.4M   

  Total Profit (millions)  
0                  $36.6M  
2                   $3.3M  
4                   $3.3M  
3                   $2.7M  
1                   $2.2M